src/algorithms/conditional_algorithm.cc

```diff
@@ -11,11 +13,26 @@ void ConditionalAlgorithm::initialize() {
   cond_mixing = std::dynamic_pointer_cast<ConditionalMixing>(mixing);
 }
 
-//! \param grid Grid of points in matrix form to evaluate the density on
-//! \param coll Collector containing the algorithm chain
-Eigen::MatrixXd ConditionalAlgorithm::eval_lpdf(
-    BaseCollector *const collector, const Eigen::MatrixXd &grid,
-    const Eigen::MatrixXd &hier_covariates /*= Eigen::MatrixXd(0, 0)*/,
-    const Eigen::MatrixXd &mix_covariates /*= Eigen::MatrixXd(0, 0)*/) {
-  return Eigen::MatrixXd::Zero(1, 1);  // TODO
+Eigen::VectorXd ConditionalAlgorithm::lpdf_from_state(
+    const Eigen::MatrixXd &grid, const Eigen::MatrixXd &hier_covariates,
+    const Eigen::MatrixXd &mix_covariates) {
+  Eigen::VectorXd lpdf(grid.rows());
+  unsigned int n_data = curr_state.cluster_allocs_size();
+  unsigned int n_clust = curr_state.cluster_states_size();
+  cond_mixing->set_state_from_proto(curr_state.mixing_state());
+  Eigen::MatrixXd lpdf_local(grid.rows(), n_clust);
+  auto temp_hier = unique_values[0]->clone();
+  for (size_t j = 0; j < n_clust; j++) {
+    Eigen::VectorXd logweights =
+        cond_mixing->get_weights(true, false, mix_covariates.row(j));  // TODO
```

Comment:
I had forgot `mixing->set_state_from_proto()` existed. With that, recovering weights from a chain state is really easy. Therefore, what we were discussing earlier is sorted out now.

However, in the last line, `mix_covariates.row(j)` I kind of wrote randomly, since I don't understand which covariate(s) are we supposed to pass to this function. This is also true for the MarginalAlgorithm variant of this function: right now we don't pass any covariates to cluster mass functions. What even is this `mix_covariates` grid of mixing covariates anyway? How many rows does it have?

Reply:
Ok so this method should return the lpdf evaluated on a grid for a single observation (ie the predictive density)
So here there should be only one mix_covariate and one hier_covariate. I think we can also have one of each in the lpdf_grid method

Reply:
Right. An `lpdf_from_state()` for single values would be either kind of a mess, or a huge impact on performance, so I kept everything in this one function. It's just a nested loop anyway. Let me know if you like it. I also went and changed the marginal version, which is nearly identical.